File: 2018/day06-1.ts
type Coords = {
  x: number;
  y: number;
};

function solution(content: string) {
  const pivots = content.split("\n").map(parse);
  const hlimit = Math.max(...pivots.map(({ x }) => x));
  const vlimit = Math.max(...pivots.map(({ y }) => y));
  const map = Array.from({ length: hlimit + 1 })
    .map((_, x) =>
      Array.from({ length: vlimit + 1 }).map((_, y) =>
        nearest(pivots, { x, y })
      )
    )
    .flat();
  const borders = borderlines(map, vlimit);
  const noBord = map.filter((c) => !!c && !includes(c, borders));
  return Math.max(
    ...Array.from(new Set(noBord)).map(
      (f) => noBord.filter((c) => c?.x === f?.x && c?.y === f?.y).length
    )
  );
}

function includes(c: Coords, coords: Set<Coords>) {
  return !!Array.from(coords).find((i) => i.x === c.x && i.y === c.y);
}

function borderlines(coords: (Coords | null)[], vlimit: number) {
  return new Set(
    coords
      .slice(0, vlimit + 1)
      .concat(coords.slice(-vlimit - 1))
      .concat(coords.filter((_, index) => index % (vlimit + 1) === 0))
      .concat(
        coords.slice(vlimit).filter((_, index) => index % (vlimit + 1) === 0)
      )
      .filter((c) => !!c)
  );
}

function nearest(pivots: Coords[], c: Coords) {
  const distances = pivots.map((p) => manhattan(p, c));
  const minDist = Math.min(...distances);
  if (distances.filter((item) => item === minDist).length > 1) {
    return null;
  }
  return pivots[distances.indexOf(minDist)];
}

function manhattan(c1: Coords, c2: Coords) {
  return Math.abs(c1.x - c2.x) + Math.abs(c1.y - c2.y);
}

function parse(line: string): Coords {
  const [x, y] = line.split(", ").map(Number);
  return { x, y };
}

Deno.readTextFile("day06.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2018/day06-1.hs
import Data.List (elemIndex)
import Data.Maybe (catMaybes, fromJust, fromMaybe, isJust)
import Data.Set (fromList, toList)

type Coords = (Int, Int)

solution :: String -> Int
solution content = maximum . map (\f -> length . filter (== f) $ noBord) $ dedup noBord
  where
    pivots = map parse . lines $ content
    hlimit = maximum . map fst $ pivots
    vlimit = maximum . map snd $ pivots
    theMap = [nearest (x, y) pivots | x <- [0 .. hlimit], y <- [0 .. vlimit]]
    borders = dedup . borderlines vlimit $ theMap
    noBord = filter (`notElem` borders) $ catMaybes theMap

dedup :: (Ord a) => [a] -> [a]
dedup = toList . fromList

borderlines :: Int -> [Maybe Coords] -> [Coords]
borderlines vlimit xs = catMaybes (one ++ two ++ thr ++ fou)
  where
    one = take (vlimit + 1) xs
    two = take (vlimit + 1) (reverse xs)
    thr = extract (vlimit + 1) xs
    fou = extract (vlimit + 1) (drop 1 xs)

extract :: Int -> [a] -> [a]
extract n = helper (n - 1)
  where
    helper _ [] = []
    helper 0 (x : xs) = x : helper (n - 1) xs
    helper m (x : xs) = helper (m - 1) xs

nearest :: Coords -> [Coords] -> Maybe Coords
nearest c ps
  | (length . filter (== minDist) $ distances) > 1 = Nothing
  | otherwise = Just ((!! (fromMaybe 0 . elemIndex minDist $ distances)) ps)
  where
    distances = map (manhattan c) ps
    minDist = minimum distances

manhattan :: Coords -> Coords -> Int
manhattan (c1x, c1y) (c2x, c2y) = abs (c1x - c2x) + abs (c1y - c2y)

parse :: String -> Coords
parse xs =
  ( read . init . head . words $ xs,
    read . last . words $ xs
  )

main :: IO ()
main = do readFile "day06.dat" >>= print . solution
